man/man1/ccdmath.1
.TH CCDMATH 1NEMO "17 August 2023"

.SH "NAME"
ccdmath \- map arithmetic using function expressions

.SH "SYNOPSIS"
.PP
\fBccdmath in=\fP[image1[,image2,...]]  \fBout=\fPimage \fBfie=\fPexpression  

.SH "DESCRIPTION"
\fBccdmath\fP combines maps through some simple operations to make a new map.
It also handles a few arithmetic functions, such as sqrt, exp, log10. See
\fIfie(3NEMO)\fP for a full definition of the available functions and
operations. Input maps are referred to by \fB%n\fP, where \fBn\fP
is the ordinal number of the input maps (1 being the first one), 
in the function expression \fBfie\fP. For example %1+%2 would add the first
two maps.
.PP
The program can also be used in creation mode: when no input files are
given, parameters \fB%x\fP, \fB%y\fP and \fB%z\fP in \fBfie=\fP 
can be used to refer to the X, 
Y and Z pixel coordinates in the map to be created, (0,0,0) 
in pixel coordinates w.r.t. reference pixel.
\fB%w\fP and \fB%r\fP can be used for 2D and 3D radius w.r.t. reference pixel, again
in pixel coordinates.

.SH "PARAMETERS"
The following parameters are recognized in any order if the keyword is also
given:
.TP 20
\fBin=\fIin-file(s)\fP
list of input file name(s), each in \fIimage(5NEMO)\fP format.
In case of multiple files, the list must be space or comma separated.
If no files are given, the program is in creation mode.
.TP
\fBout=\fIout-file\fP
output file, will be in \fIImage(5NEMO)\fP format. The image header (WCS etc.) will
be inherited from the first map on the input list.
[no default]
.TP
\fBfie=\fIexpression\fP
Expression how to manipulate the input maps or
create output map. Input maps are referred to by \fB%n\fP or \fB$n\fP,
in creation mode position is referred to by \fB%x\fP and \fB%y\fP
[no default].
.TP
\fBsize=\fInx,ny,nz\fP
Size of the output map in case program is in creation mode [default: 10,10,1].
.TP
\fBcrpix=\fIr1,r2,r3\fP
(New) values for the reference pixel. Note these are in FITS notation,
i.e. (1,1,1) is the lower left of the first plane in a cube. Recall
that (nx+1)/2 is the central pixel in coordinate X, not nx/2!
By default the WCS is transferred from the first input image to the output
image, however, by using all 3 keywords 
\fBcrpix=,crval=,cdelt=\fP the WCS of the output image can be overwritten.
[Default: \fB1,1,1\fP for image creation].
.TP
\fBcrval=\fIr1,r2,r3\fP
(New) values for the reference pixel.
[Default: \fB0,0,0\fP for image creation].
.TP
\fBcdelt=\fIr1,r2,r3\fP
(New) values for the pixel separation. 
[Default: \fI1,1,1\fP for image creation].
.TP
\fBseed=\fP\fIinteger\fP
seed for the random number generator (default: a value 0, which will
be converted into a unique new value using UNIX's clock time, in
seconds since once-upon-a-time-in-the-seventies).
.TP
\fBreplicate=t|f\fB
Normally each input image needs to be of the same shape. Setting \fBreplicate=t\fP
will allow the 3rd axis of a 2D map to be replicated in order for the \fBfie=\fP
expression to be parsed. Default: false

.SH "EXAMPLE"
Create a 'difference' map from two input maps:
.nf
   % \fBccdmath inmap1,inmap2 outmap "abs(%1-%2)"\fP
.fi
create the maximum of two maps:
.nf
   % \fBccdmath inmap1,inmap2 outmap "ifgt(%1,%2,%1,%2)"\fP
.fi
create the difference of two maps if either of the two is not zero:
.nf
   % \fBccdmath inmap1,inmap2 outmap "ifne(%1,0,%1-%2,ifne(%2,0,%1-%2,0))"\fP
.fi
create a map from scratch:
.nf
  % \fBccdmath out=newmap fie="exp(-((%x-50)/25)**2-((%y-50)/25)**2)" size=100,100\fP
.fi
create a (2 plane 4x3 image) cube from scratch:
.nf
  % \fBccdmath "" cube432 "%x+10*%y+100*%z" 4,3,2\fP
.fi
create a new WCS for an existing cube, with reference value (0,0,0) in the center of the cube,
assuming you know the size of the cube (nx,ny,nz) and pixel sizes (dx,dy,dz):
.nf
  % \fBccdmath cube0 cube1 %1 crpix=$nx/2+0.5,$ny/2+0.5,$nz/2+0.5 crval=0,0,0 cdelt=$dx,$dy,$dz\fP
.fi
and creating an annulus:
.nf
  %  \fBccdmath "" map0 "ifgt(%w,20,1,iflt(%w,10,1,0))" size=512 crpix=256,256\fP
.fi
Rescaling a map to a linear-log, retaining linear near 0, and logarithmic in signal/noise for large values
from the mean:
.nf
  set s = 0.19
  cddmath map0 map0s "sign(%1)*$s*log(1+abs(%1)/$s)"
.fi
A test \fBcube234.fits\fP with value x + 10y + 100z can be made as follows:
.nf
  ccdmath out=- fie=%x+10*%y+100*%z size=2,3,4 | ccdfits - cube234.fits
.fi

.SH "TIMING"
Performance on Rigel (SUN 3/160, 16Mhz):

.nf
.ta +1i +2i +2i
100 * 100	rang: 5.8"	+: 3.4"  	sqrt: 2.6"
200 * 200	rang: 23.2	+: 13.0"	sqrt: 8.7"
300 * 300	rang: 51.8"	+: 31.3"	sqrt: 23.8"
512 * 512	rang: 152.7"	+: 88.4"	sqrt: 64.2"
.fi

For some historic amusement, this
512^2 \fIrang\fP benchmark measured in 1989  took 153s. On a 2021 laptop
this same benchmark took 0.81/64". Scaling this it means 12,000 times
faster, in 32 years of evolution of hardware. This was a 4.2GHz i5-1135G7.
For \fIhackcode1(1NEMO)\fP this was a factor of 26,000, see \fIbench(5NEMO)\fP.

.SH "SEE ALSO"
fie(3NEMO), image(5NEMO), tsf(1NEMO), ccdgen(1NEMO), ccdsky(1NEMO)

.SH "AUTHOR"
Peter Teuben

.SH "FILES"
.nf
.ta +2.5i
src/image/trans  	ccdmath.c ccdmath.1
.fi

.SH "UPDATE HISTORY"
.nf
.ta +1.5i +5.5i
 1-Jul-87	V1.0: Created	PJT
 6-jul-87	V1.1: order keywords changed for future enhancements	PJT
 1-jun-88	V2.0: new filestruct, although code is same         	PJT
18-dec-88	V2.1: new keyword structure, fie() used.         	PJT
22-jan-89	V2.3: can make map from scratch using %x, %y and %z	PJT
1-mar-03	V3.0: set/change the WCS				PJT
19-jun-03	V3.1: allow %w and %r, and use offset from crpix	PJT
25-aug-04	V3.2: fixed error in setting crpix (off by 2!)		PJT
25-dec-2020	V3.3: add replicate=	PJT
.fi



src/image/trans/ccdmath.c
/* 
 * CCDMATH: manipulate a series of maps to generate a new map,
 *		can also start from scratch
 *
 *	1-Jul-87	Original version PJT
 *	3-Jul-87	V1.1: order of keywords changed for future expansion
 *	1-jun-88	V2.0: renamed programname from combine to ccdmath
 *				new filestruct, although code still same
 *      18-dec-88	V2.1: Gipsy's fie() functions: order and name
 *                      of keywords adapted to fie()-stuff
 *	11-jan-89	V2.2: be able to create maps from scratch
 *                            needs the size=nx,ny keyword in this case
 *	22-jan-89	V2.3a: map can also be full 3D
 *	19-jun-89	V2.3b: fie -> fien for F2C interface
 *	 7-jul-89	V2.3c: nemoinpi instead of nemoinp, back to fie again
 *	13-nov-90       V2.4 get_nan is now in library in NEMO V2.x
 *			     also forget to return value in fie_remap
 *	 7-mar-92	V2.5 modern style - gcc 2.0 happy		PJT
 *				fie() is now in real, not float
 *	24-feb-98	V2.6 added seed=				PJT
 *	 8-sep-01	a    init_xrandom
 *      24-jul-02       2.7  allow nz=0, so it does not make a 3rd axis  PJT
 *       1-mar-03       3.0  new keywords to set/change WCS              PJT 
 *      19-jun-03       3.1  allow %w, %r for 2d/3d radii; allow crpix offset     PJT
 *      26-aug-04       3.2  fix bad error in setting crpix for cube generation   PJT
 *      10-may-05       3.2a use the wcs routines that have moved to wcsio.c      PJT
 *
 *       because of the float/real conversions and
 *       to eliminate excessive memory usage, operations 'fie' are
 *       done on a column by column basis.
 *                      
 */

#include <stdinc.h>
#include <ctype.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <strlib.h>
#include <image.h>

string defv[] = {
  "in=\n           Input file(s), separated by comma's (optional)",
  "out=???\n       Output file",
  "fie=???\n       Expression %1,%2,.. for input maps; %x,%y,%z, %w,%r for new map/cube",
  "size=10,10,1\n  2- or 3D dimensions of map/cube",
  "crpix=\n        Override/Set crpix (1,1,1)",
  "crval=\n        Override/Set crval (0,0,0)",
  "cdelt=\n        Override/Set cdelt (1,1,1)",
  "seed=0\n        Random seed",
  "VERSION=3.2b\n  24-nov-2019 PJT",
  NULL,
};

string usage = "image arithmetic and wcs";

string cvsid="$Id$";

#ifndef HUGE
# define HUGE 1.0e20
#endif

#define MAXIMAGE 20

imageptr iptr[MAXIMAGE];	/* pointers to (input) images */
int      nimage;                /* actual number of input images */
bool     mapgen = FALSE;	/* no input files: create from scratch ? */

#define MAXNAX 3

double crval[MAXNAX], crpix[MAXNAX], cdelt[MAXNAX];
int nwcs = 0;



local int set_axis(string var, int n, double *xvar, double defvar);
local int fie_remap(char *fie, bool map_create);
local void do_create(int nx, int ny, int nz);
local void do_combine(void);

extern  int debug_level;		/* see initparam() */
extern    void    dmpfien();
extern    int     inifien();
extern    void    dofien();

extern string *burststring(string,string);


void nemo_main ()
{
    string *fnames, fie;
    stream  instr[MAXIMAGE];            /* input files */
    stream  outstr;                     /* output file */
    int     size[3], nx, ny, nz;        /* size of scratch map */
    int     noper;                      /* number of images needed from oper */

    init_xrandom(getparam("seed"));
    nwcs += set_axis(getparam("crpix"),MAXNAX,crpix,1.0);
    nwcs += set_axis(getparam("crval"),MAXNAX,crval,0.0);
    nwcs += set_axis(getparam("cdelt"),MAXNAX,cdelt,1.0);
    fie = getparam("in");
    if (fie==NULL || *fie==0)
        mapgen = TRUE;      /* no input files: create maps from scratch */
    else
        fnames = burststring(getparam("in"), ", ");  /* input file names */
    if(mapgen) {            /* map from scratch */
        dprintf(1,"Generating a map from scratch\n");
        switch (nemoinpi(getparam("size"),size,3)) {
            case 1:			/*  nx[,nx,1] */
                size[1] = size[0];
		size[2] = 1;
                break;
            case 2:			/*  nx,ny[,1] */
		size[2] = 1;
                break;
	    case 3:			/*  nx,ny,nz  */
                break;
            case 0:			/*  [10,10,1] */
                dprintf(0,"Cannot have no size, default 10 assumed\n");
                size[0] = size[1] = 10;
                size[2] = 1;
                break;
            default:			/* --- some error --- */
                error("Syntax error in size keyword\n");
        }
        nx = size[0];
        ny = size[1];
        nz = size[2];
    }
    fie = scopy(getparam("fie"));
    if (fie_remap(fie,mapgen) < 0)      /* remap %x,%y,%z to %1,%2,%3 */
        error("syntax error in fie = %s",fie);
    noper = inifien(fie);                   /* highest parameter # needed */
    if (debug_level >= 5)  dmpfien();        /* debug output from inifie */
    if (noper < 0)
         error ("Error in parsing fie expression %s\n inifie returned %d",
                 getparam("fie"),noper);

    outstr = stropen (getparam("out"),"w");  /* open output file first ... */

    nimage = 0;         /* count number of images/files in in= keyword */
    while (!mapgen){         /* .. then open input files one by one */
        if (nimage==MAXIMAGE)
            error("Too many input files, maximum is %d\n",MAXIMAGE);
        if (fnames[nimage]==NULL)
            break;                      /* done with file names */
        instr[nimage] = stropen(fnames[nimage],"r");    /* open file */
        iptr[nimage] = NULL;        /* make sure to init it right */
        read_image (instr[nimage], &iptr[nimage]);
        dprintf (2,"Image %d read in\n",nimage);
        if (nimage) {                   /* check size consistency */
            if (Nx(iptr[nimage]) != Nx(iptr[nimage-1]))
                error ("Input map %d does have different Nx\n",nimage);
            if (Ny(iptr[nimage]) != Ny(iptr[nimage-1]))
                error ("Input map %d does have different Ny\n",nimage);
            if (Nz(iptr[nimage]) != Nz(iptr[nimage-1]))
                error ("Input map %d does have different Nz\n",nimage);
        }
        strclose(instr[nimage]);        /* close input file */
        nimage++;
    }
    if (!mapgen) {
        dprintf (1,"%d image files read\n",nimage);
        if (nimage < noper)
            error("Not enough input maps (%d) for %d operations\n",
                nimage, noper);
    }

    if (mapgen)
        do_create(nx,ny,nz);
    else
        do_combine();

    write_image (outstr,iptr[0]);         /* write image to file */
    strclose(outstr);
}


local int set_axis(string var, int n, double *xvar, double defvar)
{
  int i, nret;
  if (var == 0 || *var == 0) {
    for (i=0; i<n; i++) 
      xvar[i] = defvar;
    return 0;
  } else {
    nret = nemoinpd(var,xvar,n);
    if (nret < 0) error("Parsing error %d in %s",nret,var);
    for (i=nret; i<n; i++)
      xvar[i] = defvar;
    return 1;
  }
  return 999;
}



/* 
 *  Checks if string 'fie' has nothing but numbers
 *  after a %, i.e. %1, %2... when map_create==false
 *  In case map_create==true % can only be followed by
 *  an 'x', 'y' or 'z' - in this case x -> 1, y -> 2, z -> 3
 *  'w' is sqrt(x*x+y*y),   'r' is sqrt(x*x+y*y+z*z).
 *  Returns 0 on success, -1 on some failure
 */

local int fie_remap(char *fie, bool map_create)
{
    for(;*fie;fie++) {
        fie = strpbrk(fie,"%$");                /* look for a $ or % */
        if (fie==NULL)
            return(0);                          /* done if End of String */
        fie++;
        if (*fie=='x' && map_create)            /* x */
            *fie = '1';
        else if (*fie=='y' && map_create)       /* y */
            *fie = '2';
        else if (*fie=='z' && map_create)       /* z */
            *fie = '3';
        else if (*fie=='w' && map_create)       /* w (2d) */
            *fie = '4';
        else if (*fie=='r' && map_create)       /* r (3D) */
            *fie = '5';
        else if (isdigit(*fie) && !map_create)
            ;                                   /* ignore */
        else {
            dprintf(1,"Warning: syntax error starting at fie=%s\n",--fie);
            return(-1);
        }
    }
    return 0;
}

/*
 *  create new map from scratch, using %x and %y as position parameters 
 *		0..nx-1 and 0..ny-1
 */
local void do_create(int nx, int ny,int nz)
{
    double m_min, m_max, total;
    real   fin[5], fout;
    int    ix, iy, iz;
    int    badvalues;
    
    m_min = HUGE; m_max = -HUGE;
    total = 0.0;		/* count total intensity in new map */
    badvalues = 0;		/* count number of bad operations */

    if (nz > 0) {
      if (!create_cube (&iptr[0], nx, ny, nz))	/* create default empty image */
        error("Could not create 3D image from scratch");
      wcs_f2i(3,crpix,crval,cdelt,iptr[0]);

      for (iz=0; iz<nz; iz++) {
        fin[2] = iz-crpix[2]+1;   /* crpix is 1 for first pixel (FITS convention) */
        for (iy=0; iy<ny; iy++) {
	  fin[1] = iy-crpix[1]+1;
	  for (ix=0; ix<nx; ix++) {
	    fin[0] = ix-crpix[0]+1;
	    fin[3] = sqrt(sqr(fin[0])+sqr(fin[1]));              /* w */
	    fin[4] = sqrt(sqr(fin[0])+sqr(fin[1])+sqr(fin[2]));  /* r */
	    dofien (fin, 1, &fout, 0.0);   /* do the work --- see: fie.3 */
	    CubeValue(iptr[0],ix,iy,iz) = fout;
	    m_min = MIN(m_min,fout);         /* and check for new minmax */
	    m_max = MAX(m_max,fout);
	    total += fout;                   /* add up totals */
	  }
        }
      }
    } else {
      if (!create_image (&iptr[0], nx, ny))	
        error("Could not create 2D image from scratch");
      wcs_f2i(2,crpix,crval,cdelt,iptr[0]);

      for (iy=0; iy<ny; iy++) {
	fin[1] = iy;
	for (ix=0; ix<nx; ix++) {
	  fin[0] = ix;
	  dofien (fin, 1, &fout, 0.0);   /* do the work --- see: fie.3 */
	  MapValue(iptr[0],ix,iy) = fout;
	  m_min = MIN(m_min,fout);         /* and check for new minmax */
	  m_max = MAX(m_max,fout);
	  total += fout;                   /* add up totals */
	}
      }
    } 
    
    MapMin(iptr[0]) = m_min;
    MapMax(iptr[0]) = m_max;

    dprintf(1,"New min and max in map are: %f %f\n",m_min,m_max);
    dprintf(1,"New total brightness/mass is %f\n",
			total*Dx(iptr[0])*Dy(iptr[0]));
    if (badvalues)
    	warning ("There were %d bad operations in dofie",badvalues);
}

/* 
 *  combine input maps into an output map  -- still 2D only
 */
local void do_combine()
{
    double m_min, m_max, total;
    real  *fin, *fout;
    int    k, ix, iy, iz, nx, ny, nz, offset;
    int    badvalues;
    
    m_min = HUGE; m_max = -HUGE;
    total = 0.0;		/* count total intensity in new map */
    badvalues = 0;		/* count number of bad operations */

    nx = Nx(iptr[0]);
    ny = Ny(iptr[0]);
    nz = Nz(iptr[0]);
    if (nwcs) {
      if (nwcs==3) 
	wcs_f2i(2,crpix,crval,cdelt,iptr[0]);
      else
	warning("Not enough WCS information given (%d/3 keywords) to replace it",nwcs);
    }

    fin = (real *) allocate(nimage*ny*sizeof(real)); 
    fout = (real *) allocate(ny*sizeof(real));
        
    for (iz=0; iz<nz; iz++)
    for (ix=0; ix<nx; ix++) {
        for (k=0; k<nimage; k++) {       /* prepare input column buffer */
            offset = ny*k;
            for (iy=0; iy<ny; iy++)
                fin[iy+offset] = CubeValue(iptr[k],ix,iy,iz);
        }
        dofien (fin, ny, fout, 0.0); /* do the work --- see: nemofie.3 */
        for (iy=0; iy<ny; iy++) {             /* write buffer back to map-0 */
            CubeValue(iptr[0],ix,iy,iz) = (real) fout[iy];
            m_min = MIN(m_min,fout[iy]);         /* and check for new minmax */
            m_max = MAX(m_max,fout[iy]);
            total += fout[iy];
        }
    }
    free(fin);
    free(fout);    

    MapMin(iptr[0]) = m_min;
    MapMax(iptr[0]) = m_max;

    dprintf(1,"New min and max in map are: %f %f\n",m_min,m_max);
    dprintf(1,"New total brightness/mass is %f\n",
			total*Dx(iptr[0])*Dy(iptr[0]));
    if (badvalues)
    	warning("There were %d bad operations in dofie",badvalues);
    
}

